[1단계 - 출석부] 히포 미션 제출합니다.

kwonkeonhyeong opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->

## 체크 리스트

- [ ] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  - <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->

[프롤로그](https://prolog.techcourse.co.kr/studylogs/3969)


## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.

<!-- 이유 작성 -->

TDD를 이용해서 구현을 해보도록 최대한 노력을 하였지만 red, green, refator의 순서를 잘 지키는 것에 대해서 어려움이 있었습니다.

이외에 리펙터링을 하는 객체의 역할과 책임을 최대한 적절하게 분배 해보려고 노력하였지만, 이 과정에서 놓친 생활 체조 원칙이 많아 3점을 주었습니다

  
## 어떤 부분에 집중하여 리뷰해야 할까요?

<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 피드백받고 싶은 내용을 공유해주세요.  
예를 들어, 가장 고민했던 점이나 여전히 어려운 부분, 그리고 이에 대한 생각을 적을 수 있습니다. -->

이번 미션에서 채택한 방법은 출석부에서 요구하는 최소한 기능을 파악하고, 해당 기능부터 TDD를 시작하는 방법이었습니다.

먼저 저희가 파악한 출석부의 최소한의 기능은 **"크루의 출석 정보를 저장, 수정, 조회하고 통계를 제공한다"** 였습니다.

그래서 해당 기능 내용을 바탕으로 출석 정보를 저장하는 테스트부터 진행 -> 구현 -> 리펙터링 하는 순서로 구현해 나가며

각각의 도메인에 대한 단위테스트는 추후 리펙터링에 의해서 도메인들이 세분화될 때 진행하고자 하였습니다.

여기서 궁금한 점은 **TDD는 어느 깊이에서 부터 시작해야 할까??** 입니다!

위에서 말씀드린 것과 같이 이번 미션은 최소한 기능을 이라고 파악된 부분부터 시작하였는데, 이와 다르게 더 깊이 들어가 Crew 같이 세부 도메인에 대한 검증과 관련된 부분부터 단위테스트를 작성하면서 TDD를 시작하는 방법이 더 효율적인지 궁금증이 생겨서 질문을 드렸습니다!

그리고 두번째로 궁금한 점은 red, green, refactor의 과정에서 red 테스트가 컴파일에러가 나서 레드가 되는 실패 테스트를 작성하고 이를 구현 리펙터링하는 것인지? 아니면 Assertion까지 작성하여 컴파일은 되지만 테스트만 실패하는 상태를 레드라고 하는 것인지가 궁금합니다!

이번 미션에서는 컴파일 에러가 발생해서 실패가 되는 테스트를 red라고 가정하고 TDD 사이클을 진행해 보았는데 red,green,refactor를 잘 연습해 본 것인지가 의문이 들었습니다! 

---

이외에 최대한 TDD를 연습해보고자 미션을 진행하다 보니 구현을 완료하지 못한 부분이나 코드가 지저분한 점이 있어 해당 부분은 미리 죄송하다는 말씀 드립니다 😭

TDD 관점에서 따끔한 리뷰 부탁드립니다!

혹시 위에 질문 중 이해가 안되는 부분이 있으시면 말씀해주세요!🙌

cookie-meringue committed:
remove: 필요 없는 .gitkeep 제거

cookie-meringue committed:
feat: Application 생성

cookie-meringue committed:
docs: 구현 기능 목록 작성

cookie-meringue committed:
docs: 출석 기록 csv 파일 추가

cookie-meringue committed:
test(red): 크루 출석 확인 기능 red 테스트 작성

cookie-meringue committed:
feat(green): 크루 출석 확인 기능 구현

- 출석부, 크루 클래스 구현.

cookie-meringue committed:
docs: 크루 출석 확인 기능 완료 표시

cookie-meringue committed:
test: 출석 기록 확인 테스트 작성

cookie-meringue committed:
docs: 출석 기록 확인 기능 완료 표시

cookie-meringue committed:
test(red): 이미 출석한 경우 예외 발생 red 테스트 작성

cookie-meringue committed:
refactor: value 변수명 복수형으로 변경

cookie-meringue committed:
feat(green): 이미 출석한 경우 예외 발생 기능 구현

cookie-meringue committed:
docs: 이미 출석한 경우 예외 발생 완료 표시

cookie-meringue committed:
refactor(test): 출석부 생성 로직 함수 분리

cookie-meringue committed:
docs: 출석 수정 부가 기능 작성(추후 구현)

cookie-meringue committed:
test(red): 출석 수정 기능 red 테스트 작성

cookie-meringue committed:
feat(green): 출석 수정 기능 구현

cookie-meringue committed:
docs: 출석 수정 기능 구현 완료 표시

cookie-meringue committed:
refactor(test): 출석 변경 시간 변수명을 함수명과 통일

cookie-meringue committed:
test(red): 출석 기능 수정 후 출석 기록 확인 기능 red 테스트 작성

cookie-meringue committed:
feat(green): 출석 기록 수정 응답 기능 구현

cookie-meringue committed:
docs: 출석 기록 수정 응답 기능 완료 표시

cookie-meringue committed:
test(red): 닉네임을 통한 출석 기록 확인 red 테스트 작성

cookie-meringue committed:
feat(green): 닉네임을 통한 출석 기록 확인 기능 구현

cookie-meringue committed:
docs: 닉네임을 통한 출석 기록 확인 기능 완료 표시

cookie-meringue committed:
refactor(test): 누락된 Test 어노테이션 삽입

cookie-meringue committed:
test(red): 제적 위험자 파악 red 테스트 작성

cookie-meringue committed:
feat(green): 제적 위험자 파악 기능 구현

cookie-meringue committed:
docs: 제적 위험자 파악 기능 완료 표시

cookie-meringue committed:
docs: 제적 위험자 리스트 반환 추가 기능 명세

cookie-meringue committed:
test(red): 제적 위험자 리스트 정렬 red 테스트 작성

cookie-meringue committed:
feat(green): 제적 위험자 정렬 기능 구현

cookie-meringue committed:
docs: 제적 위험자 정렬 기능 완료 표시

cookie-meringue committed:
refactor: 시간 변경 전, 후 데이터를 가지는 필드 추가

cookie-meringue committed:
refactor(test): Assertions 추가

cookie-meringue committed:
feat: 출석 기록 DTO 작성

cookie-meringue committed:
refactor: 출석 기록 반환 시 AttendanceLogResponse 를 반환하도록 변경

cookie-meringue committed:
refactor(test): 출석 기록 반환 Assertions 추가

cookie-meringue committed:
refactor(test): 제적 위험자 파악, 정렬 테스트를 하나로 병합

- 동일한 코드를 가지기 때문.

cookie-meringue committed:
fix: 정렬 수식 오류 수정

cookie-meringue committed:
fix: isLate 적용 범위 수정

- 결석은 계산하지 않도록 수정.

cookie-meringue committed:
feat: 달력 기능 구현

cookie-meringue committed:
refactor(test): 제적 위험자 정렬해서 반환 Assertions 추가

cookie-meringue committed:
refactor(test): 사용하지 않는 formatter 제거

cookie-meringue committed:
test(red): 제적, 면담, 경고 상태를 검증하는 red 테스트 작성

cookie-meringue committed:
feat(green): 제적, 면담, 경고 상태를 반환하는 로직 구현

cookie-meringue committed:
refactor(test): 전날까지의 크루 출석 기록 확인 Assertions 추가

cookie-meringue committed:
feat: TimeLogResponse DTO 작성

cookie-meringue committed:
feat: 출석 상태 Enum 구현

cookie-meringue committed:
refactor: MONDAY를 WEEKDAY와 병합

cookie-meringue committed:
feat: 결석 날짜 목록 반환 기능 구현

cookie-meringue committed:
refactor: 출결 기록 반환 시 출석하지 않은 날은 결석으로 반환하도록 변경

cookie-meringue committed:
refactor: 결석, 지각 횟수, 제적 위험 상태를 관리하는 도메인 변경

cookie-meringue committed:
feat: CrewAttendanceStatus 기능 구현

cookie-meringue committed:
test(red): 출석 수정 시 삭제 후 수정 red 테스트 작성

cookie-meringue committed:
feat(green): 출석 수정 시 삭제 후 수정 기능 구현

cookie-meringue committed:
chore: 기본적인 패키지 분리

cookie-meringue committed:
refactor: getter 대신 repository 의 역할을 수행하는 메서드를 사용하도록 변경

cookie-meringue committed:
refactor: Repository 역할을 더 잘 나타내는 클래스명으로 변경

cookie-meringue committed:
refactor: Repository 의 역할을 더 잘 수행하도록 함수명 및 시그니처 변경

cookie-meringue committed:
chore: model 내부 패키지 구조 변경

cookie-meringue committed:
refactor: Repository 와 직접 상호작용하는 로직을 Service 로 이전

cookie-meringue committed:
refactor: AttendanceAnalyzer 를 사용하지 않도록 변경

- AttendanceStatus 와 Calender 에게 AttendanceAnalyzer 가 가지고 있던 책임을 위임.

cookie-meringue committed:
remove: 사용하지 않는 AttendanceAnalyzer 삭제

cookie-meringue committed:
feat: 결석, 지각 횟수 vo 생성

cookie-meringue committed:
feat: 출석 수정 응답 DTO 생성

cookie-meringue committed:
feat: 출석 상태 DTO 생성

cookie-meringue committed:
feat: 출석 조회 응답 DTO 생성

cookie-meringue committed:
feat: 관리 대상 크루 응답 DTO 생성

cookie-meringue committed:
refactor: 클래스명, 함수명, 구조 변경

cookie-meringue committed:
docs: 크루 닉네임 기능 추가

cookie-meringue committed:
feat: 크루 닉네임 vo 추가

cookie-meringue committed:
refactor: 크루 닉네임을 String 대신 CrewName 을 사용하도록 변경

cookie-meringue committed:
feat: 기능 Enum 구현

cookie-meringue committed:
feat: 한국 기준 요일 Enum 구현

cookie-meringue committed:
feat: 입력 뷰 구현

cookie-meringue committed:
feat: 출력 뷰 구현

cookie-meringue committed:
feat: csv 에서 데이터를 읽어오는 역직렬화 클래스 구현

cookie-meringue committed:
feat: 크루명으로 크루를 조회하는 메서드 구현

cookie-meringue committed:
feat: Deserializer 를 받아 데이터를 초기화하도록 변경

cookie-meringue committed:
feat: Controller 와 소통하기 위한 함수 구현

cookie-meringue committed:
refactor: name 을 통한 정적 팩토리 구현

cookie-meringue committed:
feat: managementStatus 추가

cookie-meringue committed:
feat: attendanceCount 추가

cookie-meringue committed:
feat: before, after status 추가

cookie-meringue committed:
refactor(test): model 변화에 따른 테스트 코드 변경

cookie-meringue committed:
feat: Controller 구현

cookie-meringue committed:
feat: Controller 를 사용하도록 구현

jinyoungchoi95 commented on README.md:
```diff
@@ -1,3 +1,48 @@
 # java-attendance
 
-출석 미션 저장소
+# 구현 기능 목록
```
> 여기서 궁금한 점은 TDD는 어느 깊이에서 부터 시작해야 할까?? 입니다!

어느 깊이에서부터 시작해야한다는 정답은 없습니다. 사람마다 생각하는 깊이가 다른데 그건 사람마다 가지는 경험의 차이가 다르기 때문이에요.

미션의 목적은 여러가지의 경우에서 여러가지 시도를 해보고 경험을 한 뒤 히포가 생각하는 개발 철학을 가져가는 것이기 때문에 여러가지 방법을 도전해보시면 좋을 것 같아요.

참고로 저라면 요구사항과 도메인의 설계(개발 x 문서화 o)를 정의하는데 하루를 온전히 다 사용하고, 그걸 기반으로 가장 하위 깊이에 있는 도메인부터 개발을 진행할 것 같습니다.

kwonkeonhyeong replied:
먼저 좋은 리뷰 감사드립니다 😀

여러가지 경험을 하고 난 후 개발 철학을 가져가야 한다는 말씀이 인상깊네요!

개발을 공부하면서 항상 정답이 정해진 상황이 없다는 것이 때로는 어지럽고 헷갈리지만, 반대로 또 다양한 생각을 해볼 수 있게 해주는 매력인 것 같아요 😊

오리의 말씀처럼 미션에서 여러가지를 시도하고 도전해보면서 저만의 철학을 만들어 보도록 하겠습니다!

jinyoungchoi95 commented on src/test/java/attendance/AttendanceTest.java:
컴파일이 에러가 나는 경우는 예외로 둡니다.

커밋이나 tdd의 단위 기준은 컴파일을 기준으로 잡아보세요. 물론 red를 만드는 과정에서 컴파일 에러가 날 수는 있으나 red가 완성되고 커밋시점에서 컴파일 에러가 나는 것은 지양해야합니다.

커밋의 목적이 무엇인가요? 커밋으로 돌아가서 개발을 시작할 수 있어야하는데 컴파일 에러가 나는 것이 개발을 시작할 수 있는 지표가 될 수 있을까요?

kwonkeonhyeong replied:
커밋의 목적에 대해서 심도있게 생각하지 않았었는데, 좋은 피드백 감사드립니다!

커밋으로 돌아가서 개발을 시작할 수 있어야 하는데 컴파일 에러가 나는 상황이 담긴 red를 커밋 시점으로 잡는 행위는 이에 반하는 행동이었네요🥲 

커밋의 목적에 대해  생각해볼 수 있는 시간이 되었습니다! 감사합니다

jinyoungchoi95 commented on src/test/java/attendance/AttendanceTest.java:
```diff
@@ -0,0 +1,223 @@
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
+public class AttendanceTest {
```
<img width="761" alt="스크린샷 2025-02-21 오후 12 06 27" src="https://github.com/user-attachments/assets/286452f2-d589-4dde-a3ba-a622f70034c8" />

1. 테스트가 실패하고 있네요.
2. 전반적으로 단위 테스트가 하나도 보이지 않습니다. tdd를 통해서 실패하는 테스트를 먼저 만든다면 각 클래스마다 테스트가 있어야하는데 tdd가 맞을까요?

kwonkeonhyeong replied:
처음에 테스트를 시도하면서 탑다운 방식으로 진행하는 과정에서 시간이 부족하여 단위 테스트까지 세부적으로 작성하지 못한 부분이 있었습니다 😭

시간에 맞춰서 낸다고 조급하게 하다보니 리펙토링 과정에서 테스트도 다시 확인을 안했었네요,,, 😂

피드백 이후 리펙토링을 진행하였고, 단위테스트도 최대한 작성해 보았습니다!

jinyoungchoi95 commented on src/main/resources/attendances.csv:
<img width="417" alt="스크린샷 2025-02-21 오후 12 19 49" src="https://github.com/user-attachments/assets/825c488c-1abf-4a48-bc12-3c49380ae1c0" />

테스트가 실패하여 Application으로 확인하려했으나 제공된 Application도 데이터가 부족하여 테스트하기가 힘드네요. Application 데이터는 괜찮으니 실패하는 테스트를 수정하고 테스트코드부터 작성해주세요.

kwonkeonhyeong replied:
실패하는 테스트를 수정하고 전체적으로 리펙터링 한 후 과정과 단위테스트는 작성하는 과정에서 시간이 조금 걸렸습니다! 

열심히 수정하여 다시 리뷰 요청 드렸습니다! 감사합니다!

jinyoungchoi95 reviewed (changes requested):
안녕하세요 히포 😄

출석 미션의 리뷰어를 맡은 오리입니다.

실패하는 테스트가 있어 확인 부탁드려요. 테스트코드에 대한 코멘트 남겨두었으니 반영후에 리뷰요청부탁드립니다 :)

궁금하거나 고민이 되는 부분이 있으시다면 언제든 pr 코멘트 또는 dm으로 요청 부탁드립니다.
감사합니다 🙇‍♂️

kwonkeonhyeong committed:
fix: Command 제적 위험자 확인 Enum 상수 명 수정

kwonkeonhyeong committed:
fix: 지역 변수명 오타 수정

kwonkeonhyeong committed:
feat: 제적 위험 인원 출력 기능 구현

kwonkeonhyeong committed:
docs: application test csv 추가

kwonkeonhyeong committed:
refactor(test): 테스트 데이터 적용 로직 변경에 따른 테스트 리펙터링

kwonkeonhyeong committed:
refactor: 테스트 데이터 변경 및 출석 시간 데이터 포장으로 인한 전체 구현 변경

kwonkeonhyeong committed:
test: 닉네임을 입력하여 크루 출석 기록 확인 테스트 추가

kwonkeonhyeong committed:
test: 전날까지의 크루 출석 기록을 통해 제적 위험자 반환 테스트 추가

kwonkeonhyeong committed:
test: 전날까지의 크루 출석 기록을 통해 제적 위험자 반환 테스트 추가

kwonkeonhyeong committed:
refactor: 코드 컨벤션 적용 및 용도가 중복되는 메서드 제거

kwonkeonhyeong committed:
test(red): 크루 이름이 2~4글자가 아닌 경우 예외 발생 테스트 추가

kwonkeonhyeong committed:
refactor: CrewName 를 Crew 와 병합

kwonkeonhyeong committed:
test: TimeLog와 특정 날짜가 같은지 비교해주는 isSame() 테스트 추가

kwonkeonhyeong committed:
test: TimeLog 에 출석시간이 존재하지 않는 경우 예외 발생 테스트 추가

- getDateTime() 호출 시 예외 발생

kwonkeonhyeong committed:
refactor: timeLog getDateTime() 에러 메시지 수정 및 사용하지 않는 getter 제거

- getDateTime() 호출 시 예외 발생

kwonkeonhyeong committed:
test: TimeLog 내 출석 요일 및 시간에 따른 출석 상태 반환 테스트 추가

kwonkeonhyeong committed:
fix: TimeLogsTest 이름 변경

- TimeLogsTest -> TimeLogTest 로 변경

kwonkeonhyeong committed:
test: 캠퍼스 운영 시간 관련 TimeLog 생성 테스트 추가

kwonkeonhyeong committed:
refactor: 출석 등록 시간이 캠퍼스 운영 시간이 아닌 경우 예외 발생하도록 변경

kwonkeonhyeong committed:
fix: 예외 발생 메시지 수정

kwonkeonhyeong committed:
test: 주말 또는 공휴일 출석 기록 생성 시 예외 발생 테스트 추가

kwonkeonhyeong committed:
refactor: 주말 또는 공휴일 출석 기록 생성 시 예외 발생하도록 변경

kwonkeonhyeong committed:
refactor: 주말 또는 공휴일 출석 기록 생성 시 예외 발생하도록 변경

kwonkeonhyeong committed:
test: TimeLog 추가 시 같은 날 동일한 기록이 존재하는 경우 예외 발생 테스트 추가

kwonkeonhyeong committed:
refactor: TimeLog 추가 시 같은 날 동일한 기록이 존재하는 경우 예외 발생하도록 변경

- IllegalStateException -> IllegalArgumentException 으로 변경
- 예외 문구 수정

kwonkeonhyeong committed:
test: TimeLog 에 다른 TimeLogs 내의 모든 출석 기록을 등록하여 등록 여부 확인하는 테스트 추가

kwonkeonhyeong committed:
test: TimeLogs 에서 특정 TimeLog 제거 테스트 추가

kwonkeonhyeong committed:
test: TimeLogs에 특정 일의 TimeLog가 존재하는지 확인하는 테스트 추가

kwonkeonhyeong committed:
fix: TimeLogs 에서 특정 TimeLog 제거 테스트 assertion 수정

kwonkeonhyeong committed:
test: TimeLogs 기록을 기준으로 출석, 지각, 결석 횟수 반환 테스트 추가

kwonkeonhyeong committed:
test: 출석 및 지각 count를 기준으로 크루 관리 상태 확인 테스트 추가

kwonkeonhyeong committed:
test: 크루의 현재 출석 상태에 따른 관리 상태 체크 관련 테스트 추가

kwonkeonhyeong committed:
fix: 이미_출석한_경우_다시_출석할_수_없으며_수정_기능을_이용하도록_안내한다 테스트 예외 변경

kwonkeonhyeong committed:
test: LateCount 테스트 추가

kwonkeonhyeong committed:
test: AbsenceCount 테스트 추가

kwonkeonhyeong committed:
test: AttendanceCount 테스트 추가

kwonkeonhyeong committed:
test: AttendanceStatus 테스트 추가

kwonkeonhyeong committed:
test: Calender 테스트 추가

kwonkeonhyeong committed:
test: AttendanceRepository 테스트 추가

kwonkeonhyeong committed:
refactor: 매직 넘버 또는 하드 코딩 값 상수 처리

kwonkeonhyeong committed:
fix: 예외 문구 수정

jinyoungchoi95 commented on src/main/java/attendance/model/service/AttendanceService.java:
`AttendanceService`가 필요한 프로그램이 맞나요? 제가 보기엔 계층이 필요할만큼 복잡한 프로그램이 아닌 것으로 보입니다.

레벨 1에서는 객체지향, 클린코드, 테스트, 페어프로그래밍 등 배워야할 것들이 많이 있는데요. spring의 구조, layered architecture 등은 현재 미션에서 모두 제거하고 레벨 1에서만 배울 수 있는 것에 집중해보세요. 레벨 2에 가서도 이런 것들을 배울 기회가 충분히 많이 있고 레벨 1에서 배우는 내용들을 가져가는데도 시간이 부족합니다.

kwonkeonhyeong replied:
오리의 리뷰를 확인하고 미션에서 챙겨가야하는 부분에 대해서 다시 한번 생각해보게 되었습니다!

레벨 1에서 계층에 대한 요구사항이 존재하지 않았는데, 여기에 너무 얽매여서 미션을 진행하지 않았나 하는 생각이 드네요 🥲

정작 클린코드와 테스트, 객체지향적인 측면에 대해서 더 깊게 고민하지 못했던 부분도 많이 존재했던 것 같아요...

step2를 진행하면서 다시 구현하는 부분할 때에는 계층에 대한 고민보다는 TDD와 클린코드, 객체지향에 최대한 초점을 맞춰서 구현을 진행해 보도록 하겠습니다!

jinyoungchoi95 commented on src/main/java/attendance/model/domain/crew/AbsenceCount.java:
```diff
@@ -0,0 +1,55 @@
+
+  private final int value;
+
+  private AbsenceCount(int value) {
+    validate(value);
+    this.value = value;
+  }
+
+  public static AbsenceCount from(int value) {
+    return new AbsenceCount(value);
+  }
```
생성자와 동일한 static 메소드가 존재하는 이유가 무엇인가요? 생성자를 public하게 사용하면되는데 의미없는 코드량만 늘어나는 것 아닌가요? from에서 나타나는 의미가 명확하지도 않고요.

kwonkeonhyeong replied:
정적 팩토리 메서드를 적용해보는 과정에서 해당 패턴을 적용하야하는 곳에 대한 깊은 고민 없이 작성했던 것이 의미없는 코드량만 들어나는 결과를 만들게 되었습니다 😂

다시 한번 학습 해본 후에 적용해보겠습니다!

[refactor: 의미 없는 정적 팩토리 메서드 제거](https://github.com/woowacourse/java-attendance/pull/31/commits/e998fb4142eadc3ab2ea1a01a4a79c925570bb52)

jinyoungchoi95 commented on src/main/java/attendance/model/domain/crew/AbsenceCount.java:
```diff
@@ -0,0 +1,55 @@
+    return new AbsenceCount(value);
+  }
+
+  public int getPolicyAppliedAbsenceCount(LateCount lateCount) {
```
get에서 기대하는 바는 "가져오다"입니다. 해당 메소드는 가져오는 역할을 충실히 하고 있나요? 단순히 가져오기만 하는 것은 아닌 것으로 보여요.

kwonkeonhyeong replied:
메서드명을 작성할 때, 해당 메서드의 역할과 이름이 일치할 수 있도록 더 고민해서 작성해 보도록 하겠습니다!

jinyoungchoi95 commented on src/main/java/attendance/model/domain/crew/AbsenceCount.java:
```diff
@@ -0,0 +1,55 @@
+    return value;
+  }
+
+  private void validate(int value) {
+    if (value < 0) {
+      throw new IllegalArgumentException(INVALID_RANGE_MESSAGE);
+    }
+  }
```
public, private 메소드를 정렬하는 기준이 있을까요? `Crew`와 `AbsenceCount`만 보았을때도 기준이 명확해보이지 않네요.

kwonkeonhyeong replied:
오리의 코멘트를 확인하고 정렬 기준에 대해서 생각해 보았습니다!

리펙터링 과정에서 정렬 기준을 생각하여 적용해 보았고, 적용한 정렬 기준은 아래와 같습니다.

- 상위 메서드 아래로 사용된 하위 메서드들을 정렬하였습니다.
  - 만일 private 메서드가 public 메서드에서 사용되었다면 사용된 public 메서드 아래 private 메서드를 나열하는 방식을 적용해보았습니다!
 
- override 된 equals, hashCode, toString, compare 등은 클래스 맨 아래로 정렬하였습니다.

- 단순 값을 반환하는 getter의 경우 override 된 메서드 위에 위치하도록 정렬하였습니다.

위와 같이 정렬을 하면 코드 읽었을 때, 메소드 사용 순서대로 자연스럽게 읽어 내려갈 수 있지 않을까 하여 해당 정렬 방식을 사용헸습니다.

오리는 어떠한 방식으로 메소드를 정렬하시는지도 궁금합니다!

[refactor: 메서드 정렬 순서 변경](https://github.com/woowacourse/java-attendance/pull/31/commits/bf417d76937e6e7419559a37eda40da70048b754)

jinyoungchoi95 replied:
네 저도 비슷한 방식으로 정렬하는 것 같아요~

jinyoungchoi95 commented on src/test/java/attendance/model/domain/crew/AbsenceCountTest.java:
```diff
@@ -0,0 +1,21 @@
+class AbsenceCountTest {
+  @Test
+  void 결석_횟수는_음수가_될_수_없습니다() {
+    Assertions.assertThatThrownBy(() -> {
+          AbsenceCount.from(-1);
```
```suggestion
    Assertions.assertThatThrownBy(() -> AbsenceCount.from(-1))
```
`{`를 사용할 필요가 없습니다. intellij 에서는 대부분 안티패턴이나 무의미한 코드에 대해서 특정한 메시지를 남기니 의식하면서 개발해보시면 좋을 것 같아요.

kwonkeonhyeong replied:
세심한 리뷰 감사드립니다 😊

미쳐 신경쓰지 못하고 넘어갔던 부분들인데, 더 꼼꼼히 메시지를 인식하면서 개발해 보아야겠네요!

jinyoungchoi95 commented on src/test/java/attendance/model/domain/crew/AbsenceCountTest.java:
```diff
@@ -0,0 +1,21 @@
+    AbsenceCount absenceCount = AbsenceCount.from(2);
+    Assertions.assertThat(absenceCount.getPolicyAppliedAbsenceCount(lateCount)).isEqualTo(4);
+  }
+}
```
<img width="117" alt="스크린샷 2025-02-24 오후 10 24 38" src="https://github.com/user-attachments/assets/e60f22de-59ba-409b-a91e-b65229763778" />
github에서 금지표시가 발생하는 이유가 무엇일까요?

kwonkeonhyeong replied:
찾아보니 파일 마지막에 개행을 하지 않으면 no newline at a end of file 가 발생해서 해당 표시를 띄워준다는 것을 알게되었습니다!

개행을 해주지 않으면 POSIX에 근거하여 동작하는 컴파일러 등에서 EOF(End Of File)를 인지하지 못해 잘못된 동작을 할 수 있게 되기 때문에 항상 파일 마지막에는 개행을 넣어주어야 한다는 것을 인지하겠습니다!

[style: 개행 추가](https://github.com/woowacourse/java-attendance/pull/31/commits/04f99ab6d9b7db673b2a4158d845d5b02fad417b)

jinyoungchoi95 commented on src/main/java/attendance/model/domain/crew/comprator/DefaultCrewAttendanceComparator.java:
```diff
@@ -0,0 +1,18 @@
+            if (o1.getPolicyAppliedLateCount() == o2.getPolicyAppliedLateCount()) {
+                return o1.getCrew().getName().compareTo(o2.getCrew().getName());
+            }
+            return o2.getPolicyAppliedLateCount() - o1.getPolicyAppliedLateCount();
```
> indent(인덴트, 들여쓰기) depth를 2를 넘지 않도록 구현한다. 1까지만 허용한다.

kwonkeonhyeong replied:
해당 리뷰를 확인하고 indent를 줄여보는 과정에서 중첩 비교가 필요한 상황에서는 .thenComparing()을 사용하면 된다는 것을 알게 되었고 이를 적용해보았습니다.

좋은 리뷰 감사드립니다🤣

[refactor: CrewAttendanceComparator .thenComparing() 적용](https://github.com/woowacourse/java-attendance/pull/31/commits/803b8bbbd8d986e1e81a4d76de74cd8fee738f0b)

jinyoungchoi95 commented on src/main/java/attendance/model/domain/crew/comprator/CrewAttendanceComparator.java:
```diff
@@ -0,0 +1,8 @@
+import attendance.model.domain.crew.CrewAttendanceStatus;
+import java.util.Comparator;
+
+@FunctionalInterface
```
`FunctionalInterface`는 어떨때 사용하나요? 현재 사용되는 곳이 없는 것 같네요.

kwonkeonhyeong replied:
사용 했을 때의 의도는, @FunctionalInterface 애노테이션을 붙여주면 인터페이스에서 하나의 추상 메서드만 가질 수 있도록 강제하게 되어 CrewAttendanceComparator에 compare() 이외의 다른 메서드를 구현하지 못하도록 해보고자 사용했었습니다!

리뷰를 보고 찾아보니 단순히 추상메서드를 하나만 가지도록 보장하기 위해서 @FunctionalInterface를 사용하는 것이 @FunctionalInterface를 사용한 이유라고 보기에는 허점이 있네요,,

람다나 메소드 참조를 사용하여 간결하게 사용하고자 하는 목적 때문에 @FunctionalInterface를 붙이는 이유가 더 크다고 생각하게 되어 해당 애노테이션을 제거하였습니다!

[refactor:](https://github.com/woowacourse/java-attendance/pull/31/commits/105949865ff01ebcab9d8066ad9a0f535c45fec7) @FunctionalInterface [제거](https://github.com/woowacourse/java-attendance/pull/31/commits/105949865ff01ebcab9d8066ad9a0f535c45fec7)

jinyoungchoi95 commented on src/main/java/attendance/model/domain/crew/CrewAttendanceStatus.java:
```diff
@@ -0,0 +1,64 @@
+public class CrewAttendanceStatus {
+
+  private final Crew crew;
+  private final AbsenceCount absenceCount;
+  private final LateCount lateCount;
+  private final AttendanceCount attendanceCount;
```
count를 각각 나눠야하는 이유가 있을까요? `CrewAttendanceStatus`라는 클래스가 너무 복잡해보이지 않나요?

```java
private final Map<AttendanceStatus, Integer>
```
로 표현되어도 충분해보이네요.

kwonkeonhyeong replied:
굳이 나누지 않고 Map<AttendanceStatus, Integer> 로 표현하도록 로직을 변경해 보았습니다!

여기서 궁금했던 점은 Map<AttendanceStatus, Integer> 와 같은 자료구조가 필드에 존재할 때에, 무조건 일급 컬렉션으로 감싸줘야 하는 것이 궁금합니다.

리펙터링 과정에서 Map<AttendanceStatus, Integer> 를 AttendanceStatusCounts 로 포장해서 해당 객체에게 정책이 적용된 (지각이 3회면 결석 1로 친다) 결석 수를 반환해 주는 책임 등을 넘겨주었습니다. 하지만 이외에 단순히 출석, 지각, 결석 수를 꺼내 쓰는 과정에서 객체로 한번 포장이 되어있다 보니 getter를 사용해 한번 더 꺼내써야 된다는 점이 발생했고, 다른 객체에서 AttendanceStatusCounts 안에 있는 값을 사용하고자 할 때에 AttendanceStatusCounts를 필드로 가지고 있는 객체에서 getter를 한번 더 만들어 주게 되어 코드의 복잡도가 올라가지 않았나? 하는 생각이 들었습니다.

그래서 중요한 비지니스 로직을 가지게 되는 경우를 제외하고는 일급 컬렉션으로 무조건 포장하지 않는 것도 객체 설계의 복잡도를 낮추는 좋은 방법이지 않을까 생각하는데, 이에 대한 오리는 어떻게 생각하시는지 궁금합니다!

[refactor: AttendanceStatusCounts 로 출석 결과 관련 수 포장](https://github.com/woowacourse/java-attendance/pull/31/commits/39b74f7ac77f8b8439b309a7623f72cc1269cdb7)

jinyoungchoi95 commented on src/main/java/attendance/model/domain/crew/CrewAttendanceStatus.java:
```diff
@@ -0,0 +1,64 @@
+        managementStatus);
+  }
+
+  public boolean requiresManagement() {
```
boolean의 경우 true와 false에 대해서 어떤 것이냐를 묻는 것이기 때문에 요구하다를 의미하는 require보다는 더 명확한 네이밍이 필요해보여요.

kwonkeonhyeong replied:
더 정확한 네이밍 신경쓰도록 하겠습니다!

jinyoungchoi95 commented on src/main/java/attendance/model/domain/log/TimeLog.java:
가지고 있는 것은 `TimeLog`밖에 없는데 log라는 패키지 네이밍은 너무 추상화되어있지 않나요?

kwonkeonhyeong replied:
굳이 패키지로 나누지 않아도 될 것 같아서 묶여있던 패키지를 제거하였습니다!

[chore: log 패키지 제거](https://github.com/woowacourse/java-attendance/pull/31/commits/f1e37895d598367896ab7a6a1109835da337993f)

jinyoungchoi95 commented on src/main/java/attendance/model/domain/log/TimeLog.java:
```diff
@@ -0,0 +1,136 @@
+  }
+
+  public LocalDateTime getDateTime() {
+    if(time == null) {
```
```suggestion
    if (time == null) {
```

jinyoungchoi95 commented on src/main/java/attendance/model/domain/log/TimeLog.java:
```diff
@@ -0,0 +1,136 @@
+public class TimeLog implements Comparator<TimeLog> {
+
+  private static final String NON_EXISTS_TIME_LOG = "출석 시간이 존재하지 않습니다";
+  private static final String NON_BUSINESS_HOURS_MESSAGE = "캠퍼스 운영시간이 아닙니다 (운영시간 매일 08:00~23:00)";
```
운영시간의 코드가 변경되면 에러 메시지도 함께 변경해주어야하는 불편함이 있네요. 이미 상수로 정의된 `LocalTime`을 사용해볼까요?

kwonkeonhyeong replied:
상수도 변경 가능성에 유연하게 대처할 수 있도록 신경 써야 한다는 점을 배웠습니다!
감사합니다!

[refactor: 운영시간 코드 에러 메시지 내 운영시간 LocalTime 활용](https://github.com/woowacourse/java-attendance/pull/31/commits/44e97cdfec303bee3cd44396f72b69f06e0a8c31)

jinyoungchoi95 commented on src/main/java/attendance/model/domain/log/TimeLog.java:
```diff
@@ -0,0 +1,136 @@
+import java.util.Comparator;
+import java.util.Objects;
+
+public class TimeLog implements Comparator<TimeLog> {
```
null을 사용하여 로직이 구성되는 경우가 많은데 모든 null을 제거하고 도메인을 구성해보세요. final로 불변을 챙긴다고해도 null을 사용하면 객체의 안정성 보장이 되지 못합니다.

kwonkeonhyeong replied:
null 사용했던 이유는 csv 파일을 읽어 출석 데이터를 초기화할 때 출석 기록이 없는 경우는 시간을 null로 채워서 시간이 null인 경우를  결석으로 처리하기  위함이었습니다.

오리의 리뷰를 확인하고 생각해 보았을떄, 코드에서 nullpointerexception을 막기위해서 time == null 등의 로직을 사용하는 것 부터가 객체가 안정성을 보장하고 있지 않는다 나타내고 있다고 생각하게 되었습니다!

그래서 기존에 출석 데이터를 초기화할 때 출석 기록이 없는 경우에 시간을 null로 채우던 로직을 제거하고 실제로 .csv 파일에 존재하는 데이터만 초기화하여 timelog에 null이 저장되어 사용되는 상황을 걷어내었습니다.

다만 추후 출석하지 않은 날을 view에서 알아야 이를 확인하고 "--:--" 으로 보여줄 수 있기 때문에 view에게 전달하는 AttendanceInformation의 time에 null을 채워서 view 에게 전달하고 view 에서 비어있는 시간을 판단해서 보여주는 방식을 사용했습니다.

여기서 궁금했던 점은 단순히 값만 옮기는 dto 역할을 하는 객체에게는 비어있는 데이터에 대해서 null을 담고 view에서 이를 판단는 로직 또한 도메인 안정성을 해치는 행위인지 궁금합니다!

[refactor: TimeLog 의 null 관련 로직 제거](https://github.com/woowacourse/java-attendance/pull/31/commits/83cde385c4f31070ee3f3d1ff01826ac829a9c53)

jinyoungchoi95 replied:
null 자체가 없도록 해보세요. 추상화를 활용하면 null이 없도록도 가능해보입니다.
```java
public interface A {
}

public class WithAge implements A {
    private final int age;
    private final String name;
}

public class NoAge implements A {
    private final String name;
}
```

jinyoungchoi95 commented on src/main/java/attendance/model/repository/CrewAttendanceDeserializer.java:
`Deserializer`가 무엇인가요? 이 객체는 `Deserializer`가 맞나요?

kwonkeonhyeong replied:
Deserializer의 의미를 제대로 알지 못하고 사용했습니다! 

추상화 과정에서 네이밍을 변경해보았습니다!

[refactor: CrewAttendanceDeserializer 를 CrewTimeLogsInitializer 로 변경](https://github.com/woowacourse/java-attendance/pull/31/commits/e9cacd164bf8cff1932793fae3b92a1dbd521852)

jinyoungchoi95 commented on src/main/java/attendance/model/repository/CrewAttendanceDeserializer.java:
```diff
@@ -0,0 +1,51 @@
+
+  public Map<Crew, TimeLogs> readAll(final Path filePath) throws UncheckedIOException {
+    try (Stream<String> lines = Files.lines(filePath)) {
+      HashMap<Crew, TimeLogs> map = new HashMap<>();
```
해당 코드에 추상화를 적용해볼 수 있을까요?

kwonkeonhyeong replied:
해당 기능에서 어느 부분을 추출해서 추상화를 해야할 지 고민이 되었어요!

리펙토링 과정에서 csv에서 데이터를 읽어서 Map<Crew, TimeLogs>로 파싱하는 과정 전체를 추상화 해보았습니다.

이렇게 추상하게 된 이유는 파일을 읽어들이는 방법에 따라 차이가 발생할 수 있다고 생각했기 때문입니다!

[refactor: CrewAttendanceDeserializer 를 CrewTimeLogsInitializer 로 변경](https://github.com/woowacourse/java-attendance/pull/31/commits/e9cacd164bf8cff1932793fae3b92a1dbd521852)

jinyoungchoi95 replied:
```java
HashMap<Crew, TimeLogs> map = new HashMap<>();
```
로 되어있어

```java
Map<Crew, TimeLogs> map = new HashMap<>();
```
로 전환하는 것을 말씀드린 내용이었어요. 자료구조에 대한 추상화는 기본적인 것이니 추상화에 대해서 조금 더 의식해볼 필요가 있을 것 같아요.

jinyoungchoi95 commented on src/main/java/attendance/dto/AttendanceLogResponse.java:
말씀드린대로 spring에서 사용하시던 모든 용어들은 잊고 레벨 1에 집중해보세요. 자바 콘솔 게임에 Response라는 네이밍이 적절한 네이밍일까요? 명시적이고 구체적으로 객체를 나타낼 수 있는 네이밍들을 활용해보세요.

kwonkeonhyeong replied:
spring에서 사용하던 용어를 잊고 콘솔 게임에 어울리는 네이밍을 적용하도록 변경해 보았습니다!

[refactor: Response DTO 네이밍 변경](https://github.com/woowacourse/java-attendance/pull/31/commits/4260a9a9f79f79e6b4da0e2bb852aa2fc9b71a17)

jinyoungchoi95 commented on src/main/java/attendance/controller/AttendanceController.java:
```diff
@@ -0,0 +1,109 @@
+  private final AttendanceService attendanceService = new AttendanceService(attendanceRepository);
+  private final CrewAttendanceComparator crewAttendanceComparator = new DefaultCrewAttendanceComparator();
+
+  public void run() {
+    Command command;
+    do {
+      command = inputView.inputCommand();
+
+      if (command == Command.ATTENDANCE_CHECK) {
+        attendance();
+      }
+
+      if (command == Command.ATTENDANCE_MODIFY) {
+        updateAttendance();
+      }
+
+      if (command == Command.CREW_ATTENDANCE_CHECK) {
+        checkCrewAttendance();
+      }
+
+      if (command == Command.EXPULSION_CREW_CHECK) {
+        searchManagementCrews();
+      }
+
+    } while (command != Command.QUIT);
+
+  }
```
> 함수(또는 메소드)의 길이가 15라인을 넘어가지 않도록 구현한다.

kwonkeonhyeong replied:
Command 에 각 기능에 해당하는 AttendanceController의 메서드를 Consumer로 추가하여 실행하는 방식으로 리펙터링하여 메서드 길이와 로직을 변경해 보았습니다!

[refactor: Controller run() 메서드 분리](https://github.com/woowacourse/java-attendance/pull/31/commits/b1c868729399b6ccb57deb3449661575e8955a34)

jinyoungchoi95 commented on src/main/java/attendance/model/domain/crew/AbsenceCount.java:
```diff
@@ -0,0 +1,55 @@
+  }
+
+  @Override
+  public String toString() {
```
`toString()`은 어디에서 활용되어서 override되었을까요?

kwonkeonhyeong replied:
값을 저장하는 역할이 큰 객체들에 대해서  toString override은 기본적으로 해주자! 라는 생각을 가지고 있습니다!

toString을 재정의 해두면 추후 로그를 찍어본다거나 하는 과정에서 편하게 이용할 수 있지 않을까? 생각했습니다!

jinyoungchoi95 reviewed (changes requested):
안녕하세요 히포 😄

다음 단계 넘어가기 전 같이 이야기하면 좋을 내용 코멘트 남겨두었으니 확인부탁드려요.

궁금하거나 고민이 되는 부분이 있으시다면 언제든 pr 코멘트 또는 dm으로 요청 부탁드립니다.
감사합니다 🙇‍♂️

kwonkeonhyeong committed:
refactor: Controller run() 메서드 분리

kwonkeonhyeong committed:
style: 개행 추가

- no newline at a end of file 문제 해결

kwonkeonhyeong committed:
refactor: TimeLog 의 null 관련 로직 제거

- 데이터 초기화 시 출석 기록이 없는 경우에 대한 time이 null 인 기본 TimeLog 생성 로직 삭제
- CrewAttendanceLogResponse 로 반환 시에 TimeLogs 에서 출석 기록이 없는 날에 대하여 time 이 null 인 AttendanceResponse 를 생성하여 반환하도록 변경
- 데이터가 null인 경우에 대한 판단을 view 에게 위임

kwonkeonhyeong committed:
refactor: 운영시간 코드 에러 메시지 내 운영시간 LocalTime 활용

kwonkeonhyeong committed:
refactor: requiresManagement() 메서드 명 변경

- requiresManagement() -> isRequiredManagement() 로 변경

kwonkeonhyeong committed:
refactor: 의미 없는 정적 팩토리 메서드 제거

kwonkeonhyeong committed:
chore: log 패키지 제거

kwonkeonhyeong committed:
refactor: AttendanceStatusCounts 로 출석 결과 관련 수 포장

kwonkeonhyeong committed:
refactor: Response DTO 네이밍 변경

- 콘솔 게임에 적절한 이름으로 변경

kwonkeonhyeong committed:
refactor: CrewAttendanceComparator .thenComparing() 적용

kwonkeonhyeong committed:
refactor: @FunctionalInterface 제거

kwonkeonhyeong committed:
refactor: CrewAttendanceDeserializer 를 CrewTimeLogsInitializer 로 변경

- 객체 변경 및 추상화 적용

kwonkeonhyeong committed:
refactor: 메서드 정렬 순서 변경

- 상위 메서드 아래로 사용된 하위 메서드들을 나열하는 방식으로 변경
- Override 된 equals, hashCode, toString, compare 등은 클래스 맨 아래로 위치
- 단순 값을 반환하는 getter의 경우 Override 된 메서드 위에 위치

jinyoungchoi95 commented on src/main/java/attendance/model/domain/crew/AttendanceStatus.java:
```diff
@@ -0,0 +1,30 @@
+
+public enum AttendanceStatus {
+
+  ATTENDANCE("출석"),
+  LATE("지각"),
+  ABSENCE("결석");
```
화면의 요구사항이 출력, 지각, 결석이 아닌 다른 문구로 변경되면 view의 변경사항에 의해 domain 의 코드가 변경되겠네요.

kwonkeonhyeong replied:
view 의 변경 사항에 의해 영향이 받지 않도록 각각의 AttendanceStatus를 view로 직접 전달하여 view에서 Status에 따라 각자의 문구를 출력하도록 변경하였습니다 😀

[refactor: AttendanceStatus 도메인 변경 위험 사항 제거](https://github.com/woowacourse/java-attendance/pull/31/commits/79bdbce0bda966d8521f5699ca5cf3c05fb42fb1)

[refactor: ManagementStatus 도메인 변경 위험 사항 제거](https://github.com/woowacourse/java-attendance/pull/31/commits/ca505bd105ddbd74a8ab48f1e23fa319fbf8efca)

jinyoungchoi95 commented on src/main/java/attendance/model/service/AttendanceService.java:
```diff
@@ -0,0 +1,72 @@
+import java.util.List;
+import java.util.Map.Entry;
+
+public class AttendanceService {
```
Service가 아직 있는데 Service는 테스트가 필요없나요? 화면의 출력과 관련된 로직이 아닌 이상 모든 로직은 테스트대상입니다.

kwonkeonhyeong replied:
Service계층 구조를 제거하고 최대한 객체 간의 소통이 이루어 질 수 있도록 변경해 보았습니다! 

[refactor: 계층 구조 제거](https://github.com/woowacourse/java-attendance/pull/31/commits/2b8c0799733b453e1cce9a47b9ee515e8ec56130)

jinyoungchoi95 commented on src/main/java/attendance/model/domain/crew/AttendanceStatusCounts.java:
```diff
@@ -0,0 +1,47 @@
+    this.attendanceStatusCounts = attendanceStatusCounts;
+  }
+
+  public static AttendanceStatusCounts of(TimeLogs timeLogs) {
```
관습적인 네이밍을 활용하셔서 남기는 코멘트입니다.

of의 의미가 무엇인가요? 
파라메터가 1개일때는 from을 쓰는데 왜 1개일때는 from을 쓰고 여러개일때는 of를 쓸까요?

kwonkeonhyeong replied:
정적 팩터리 메서드에서 파라메터가 1개일때는 from을 사용하여 파라메터의 형식?을 변화시켜 객체를 생성할때 사용하고, 여러 요소로 부터 하나의 객체를 만들때는 of를 사용하는 것으로 알고 사용하고 있었습니다!

리펙터링 과정에서 네이밍 실수가 자주 발생했네요 😂 더 신경써야 되는 부분 리뷰해주셔서 감사드립니다!

jinyoungchoi95 commented on src/main/java/attendance/model/domain/crew/AttendanceStatusCounts.java:
```diff
@@ -0,0 +1,47 @@
+  }
+
+  public static AttendanceStatusCounts of(TimeLogs timeLogs) {
+    Map<AttendanceStatus, Integer> attendanceStatusCounts = new HashMap<>();
+    attendanceStatusCounts.put(AttendanceStatus.ATTENDANCE, timeLogs.calculateAttendanceCount());
+    attendanceStatusCounts.put(AttendanceStatus.LATE, timeLogs.calculateLateCount());
+    attendanceStatusCounts.put(AttendanceStatus.ABSENCE, timeLogs.calculateAbsenceCount());
+    return new AttendanceStatusCounts(attendanceStatusCounts);
```
`Map.of()`를 활용해볼 수 있겠네요.

kwonkeonhyeong replied:
Map.of()를 사용해서 읽기 전용으로 변경하였습니다! 좋은 리뷰 감사드립니다😀

[refactor: CrewAttendanceStatus 초기화에 Map.of() 적용](https://github.com/woowacourse/java-attendance/pull/31/commits/27a10647ba62bd7dc61209ce4b292c04af62f137)

jinyoungchoi95 commented on src/main/java/attendance/model/repository/DefaultCrewTimeLogsInitializer.java:
```diff
@@ -0,0 +1,45 @@
+      HashMap<Crew, TimeLogs> map = new HashMap<>();
+      lines.skip(1)
+          .map(this::deSerialize)
+          .forEach(entry -> {
```
stream의 foreach 종단 연산은 안티패턴입니다.

https://tecoble.techcourse.co.kr/post/2020-05-14-foreach-vs-forloop/

kwonkeonhyeong replied:
안티패턴을 제거하는 방향으로 로직을 변경하였습니다! 모르는 내용이었는데 좋은 자료 감사드립니다 👍

[refactor: 종단 연산 .foreach() 안티패턴 제거](https://github.com/woowacourse/java-attendance/pull/31/commits/f16d7266869ee11d29ecdd4367988878bdf7b6bb)

jinyoungchoi95 commented on src/main/java/attendance/model/domain/crew/TimeLog.java:
```diff
@@ -0,0 +1,130 @@
+  }
+
+  public AttendanceInformation generateAttendanceInformation() {
+    return new AttendanceInformation(date, time, getAttendanceStatus().getName());
```
getter.getter로 체이닝되는 로직들이 많은데 전반적으로 객체 중심이 아닌 데이터 중심으로 설계하여 래핑이 과해보입니다. 이전에 말씀드렸듯 객체라는 것이 무엇인지 고민해보고 객체와 객체가 서로 상호작용을 할 수 있게끔만 설계해보세요.

kwonkeonhyeong replied:
이번 리펙토링 과정에서 최대한 계층을 걷어내고 객체 중심으로 변경해 보려고 했는데, 초기 설계가 빈약하여 information쪽 래핑 로직들을 전부 걷어내지는 못 했습니다 ㅜㅜ 2단계 진행 시 참고하여 진행해 보도록 하겠습니다.

[refactor: 계층 구조 제거](https://github.com/woowacourse/java-attendance/pull/31/commits/2b8c0799733b453e1cce9a47b9ee515e8ec56130)

jinyoungchoi95 commented on src/main/java/attendance/model/domain/crew/CrewAttendanceStatus.java:
```diff
@@ -0,0 +1,53 @@
+package attendance.model.domain.crew;
+
+public class CrewAttendanceStatus {
```
단순히 래핑만하고 추가기능이 없는 클래스가 의미있는 객체인지 잘 모르겠네요.

kwonkeonhyeong replied:
단순 래핑만 하는 객체를 걷어내고 객체가 역할을 가질 수 있도록 변경해 보았습니다!

[refactor: 계층 구조 제거](https://github.com/woowacourse/java-attendance/pull/31/commits/2b8c0799733b453e1cce9a47b9ee515e8ec56130)

jinyoungchoi95 commented on src/main/java/attendance/model/domain/crew/TimeLogs.java:
```diff
@@ -0,0 +1,134 @@
+  }
+
+  public int calculateAbsenceCount() {
+    int presentAbsenceCount = Math.toIntExact(
+        logs.stream()
+            .filter(log -> log.getAttendanceStatus() == AttendanceStatus.ABSENCE)
+            .count()
+    );
+    int missingAttendanceCount = Calender.calculateMissingAttendanceDateCount(this);
+    return presentAbsenceCount + missingAttendanceCount;
```
이미 long으로 제공하는 `count()`를 굳이 int로 바꾸었어야하나라는 생각이 들어요. `calculateMissingAttendanceDateCount`도 내부 구현은 `count()`이므로 long을 반환하게 만들 수 있구요.

반드시 int여야하는 이유가 없는 것이라면 이미 제공하고 있는 기능의 타입을 쓰는 것이 더 안전해보입니다.

kwonkeonhyeong replied:
리뷰 확인 후 생각해 보았는데 딱히 int 를 사용해야 하는 이유가 없었네요 😅

보통 long을 기본으로 사용하는 것이 좋을까요?? 

물론 상황에 따라 다를 수 있겠지만 일반적으로 정수를 담을 때에는 담을 수 있는 범위가 더 넓은 long을 기본적으로 사용하는 것이 더 안전한지 궁금합니다!

[refactor: 무의미한 type 변환 제거](https://github.com/woowacourse/java-attendance/pull/31/commits/e700b3fb76e621940701d81525ae7e016d9d1211)

jinyoungchoi95 replied:
도메인 정의 시 범위가 int를 넘어가지 않는다면 굳이 기본으로 long을 활용할 필요는 없을 것 같아요.

jinyoungchoi95 commented on src/main/java/attendance/model/domain/calender/Calender.java:
```diff
@@ -0,0 +1,56 @@
+  HOLIDAY(List.of(25));
+
+  private static final int YEAR = 2024;
+  private static final Month MONTH = Month.DECEMBER;
```
요구사항에 특정 기간의 제약을 두고 보지는 않지만 `MONTH`라는 네이밍으로 어떤 의미를 하는 상수인지 알 수 있을까요? 상수의 목적은 무엇일까요?

kwonkeonhyeong replied:
의미가 있는 이름으로 확실하게 상수의 쓰임을 전달했어야 했는데 MONTH는 너무 포괄적으로 상수 명을 정했던 것 같습니다 😅

[refactor: MONTH 상수명 변경](https://github.com/woowacourse/java-attendance/pull/31/commits/4b0011a5836d3496deacf316ef3c18491b04b043)

jinyoungchoi95 commented on src/main/java/attendance/model/domain/Information/AttendanceInformation.java:
model이라는 패키지의 의미는 무엇인가요? domain과 model의 차이는 무엇인가요?

kwonkeonhyeong replied:
mvc 계층 구조를 생각하면서 패키지 구조를 설계하여 상위에 model 이라는 패키지를 두고 하위에 domain 패키지를 배치하였습니다.

리뷰를 확인 후 domain과 model의 차이에 대해서 알아보았는데, domain은 문제를 해결하기 위한 비즈니스 영역이라는 생각이 들었고, model은 이런 특정 도메인들을 세부적으로 구체화 한 것 (객체 등으로..?)이라는 생각이 들었습니다.

개념들이 헷갈리는데 오리가 생각하는 domain과 model의 차이는 무엇인지 궁금해요 🙏

jinyoungchoi95 replied:
domain은 현재 해결하고자하는 비즈니스 영역이고 model은 그걸 형태로써 만든 코드라고 볼 수 있습니다.

미션 내에서는 domain에 대해서 집중하게 될거고 model이란 용어는 사용할 필요가 딱히 없어보이네요. 현재 model로 세분화하였을 때 장점이 없을 뿐더러 해당 개념을 사용하기보다는 domain에 먼저 집중해서 생각하고 고민해보세요.

jinyoungchoi95 commented on src/test/java/attendance/model/domain/log/TimeLogTest.java:
야직 패키지가 main 코드와 다른 클래스들이 있네요

kwonkeonhyeong replied:
세심한 리뷰 감사드립니다...😂 

[chore: model 패키지 제거](https://github.com/woowacourse/java-attendance/pull/31/commits/66a7c8e4121f51644521058b206005a0f7c236ad)

jinyoungchoi95 commented on src/test/java/attendance/model/domain/log/TimeLogTest.java:
```diff
@@ -0,0 +1,162 @@
+
+class TimeLogTest {
+
+  // TimeLogs, TimeLog
```
주석은 모두 제거해주세요

jinyoungchoi95 commented on src/main/java/attendance/model/domain/crew/TimeLogs.java:
```diff
@@ -0,0 +1,134 @@
+        });
+  }
+
+  public void addAll(TimeLogs timeLogs) {
```
테스트를 위한 메소드가 존재해도 될까요?

kwonkeonhyeong replied:
리펙터링 과정에서 테스트에서만 사용되도록 남겨진 것 같습니다! 제거하였습니다

[refactor: 사용하지 않는 메서드, 불필요한 테스트 제거](https://github.com/woowacourse/java-attendance/pull/31/commits/df56a7d031a89a53f212c4dd94f9a2e88a4f4f16)

jinyoungchoi95 commented on src/test/java/attendance/model/domain/log/TimeLogsTest.java:
```diff
@@ -0,0 +1,112 @@
+
+    timeLogs.addAll(anotherTimeLogs);
+
+    assertThat(
+        timeLogs.isContain(LocalDateTime.of(2024, 12, 4, 10, 2))
+    ).isTrue();
+
+    assertThat(
+        timeLogs.isContain(LocalDateTime.of(2024, 12, 5, 10, 3))
+    ).isTrue();
```
45번라인의 assertion이 실패해도 49번라인의 테스트 결과를 피드백받을 수 있는 방법은 없을까요?

kwonkeonhyeong replied:
assertAll()을 사용해 보도록 하겠습니다!

[refactor: 개별 assertion assertAll() 적용](https://github.com/woowacourse/java-attendance/pull/31/commits/9a447e0d446de8d89f61d4867e1ba0e25dbb2ce7)

jinyoungchoi95 commented on src/test/java/attendance/model/domain/log/TimeLogsTest.java:
```diff
@@ -0,0 +1,112 @@
+  }
+
+  @Test
+  void TimeLogs에_다른_TimeLogs_내의_모든_출석_기록을_등록() {
```
> 함수(또는 메서드)의 길이가 10라인을 넘어가지 않도록 구현한다.

해당 조건은 테스트에도 적용해볼 수 있습니다. 10라인까지는 아니어도 테스트 메소드에 대한 가독성도 고려해보세요.

kwonkeonhyeong replied:
테스트에 대한 가독성을 제대로 고려하지 못했네요😅

테스트 코드도 코드다! 가독성에 더 신경쓰도록 하겠습니다

jinyoungchoi95 commented on src/test/java/attendance/model/repository/AttendanceRepositoryTest.java:
```diff
@@ -0,0 +1,131 @@
+
+class AttendanceRepositoryTest {
+
+  private AttendanceRepository attendanceRepository;
+
+  @BeforeEach
+  void beforeEach() {
+    final CrewTimeLogsInitializer crewTimeLogsInitializer = new DefaultCrewTimeLogsInitializer();
+    final Path crewAttendanceDataPath = Path.of("src/main/resources/attendances_test.csv");
+    attendanceRepository = new AttendanceRepository(crewTimeLogsInitializer,
+        crewAttendanceDataPath);
+  }
```
junit5는 기본 라이프사이클이 메소드당이기때문에 beforeEach로 가변적으로 선언할 필요는 없어요.

https://junit.org/junit5/docs/current/user-guide/#writing-tests-test-instance-lifecycle

kwonkeonhyeong replied:
내용 확인하였습니다! 감사합니다 😊

jinyoungchoi95 reviewed (changes requested):
안녕하세요 히포 😄

몇가지 코멘트 남겨두었으니 확인부탁드려요.

궁금하거나 고민이 되는 부분이 있으시다면 언제든 pr 코멘트 또는 dm으로 요청 부탁드립니다.
감사합니다 🙇‍♂️

kwonkeonhyeong committed:
chore: model 패키지 제거

kwonkeonhyeong committed:
refactor: 계층 구조 제거

kwonkeonhyeong committed:
style: 주석 제거

kwonkeonhyeong committed:
refactor: CrewAttendanceStatus 초기화에 Map.of() 적용

- HashMap을 이용한 초기화 제거
- Map.of()를 사용하여 AttendanceStatus에 대한 count 초기화
- 불변 객체 사용으로 상태 변경 방지 및 코드 간결화

kwonkeonhyeong committed:
refactor: 종단 연산 .foreach() 안티패턴 제거

kwonkeonhyeong committed:
refactor: 무의미한 type 변환 제거

- int type 사용되는 부분 long으로 변경

kwonkeonhyeong committed:
refactor: 사용하지 않는 메서드, 불필요한 테스트 제거

- TimeLogs addAll() 제거
- 관련 테스트 제거

kwonkeonhyeong committed:
refactor: 사용하지 않는 메서드 제거 및 스타일 컨벤션 적용

kwonkeonhyeong committed:
refactor: AttendanceStatus 도메인 변경 위험 사항 제거

- view의 변경사항에 따라 domain에 영향을 끼치는 코드 변경

kwonkeonhyeong committed:
refactor: ManagementStatus 도메인 변경 위험 사항 제거

- view의 변경사항에 따라 domain에 영향을 끼치는 코드 변경

kwonkeonhyeong committed:
refactor: MONTH 상수명 변경

kwonkeonhyeong committed:
refactor: 개별 assertion assertAll() 적용

kwonkeonhyeong committed:
refactor: 사용하지 않는 필드 제거

kwonkeonhyeong committed:
refactor: 불필요한 테스트 메서드 제거

jinyoungchoi95 commented on src/main/java/attendance/domain/crew/Crew.java:
```diff
@@ -0,0 +1,45 @@
+  }
+
+  private void validate(String value) {
+    if (value.length() < 2 || value.length() > 4) {
```
매직넘버를 의미있는 상수로 변경해주는 것이 좋겠네요.

jinyoungchoi95 commented on src/main/java/attendance/domain/crew/ManagementStatus.java:
```diff
@@ -0,0 +1,27 @@
+  private static final int MIN_ABSENCES_FOR_COUNSELING = 2;
+  private static final int MIN_ABSENCES_FOR_WARNING = 1;
+
+  public static ManagementStatus of(long appliedPolicyAbsenceCount) {
```
https://github.com/woowacourse/java-attendance/pull/31#discussion_r1971411899

jinyoungchoi95 commented on src/main/java/attendance/domain/crew/TimeLog.java:
```diff
@@ -0,0 +1,134 @@
+  private final LocalDate date;
+  private final LocalTime time;
+
+  private TimeLog(LocalDate date, LocalTime time) {
+    validateUnavailableTime(time);
+    validateUnavailableDate(date);
+    this.date = date;
+    this.time = time;
+  }
+
+  public static TimeLog of(LocalDate localDate, LocalTime localTime) {
+    return new TimeLog(localDate, localTime);
+  }
+
+  public static TimeLog from(LocalDateTime dateTime) {
```
생성자를 그대로 사용하여도되는데 of, from 등의 static 메소드를 추가하는 것은 코드의 양만 늘리는 것 아닌가요? 이름으로써 특별하게 전달되는 의미도 없어보이고요.

자바를 사용하는 입장에서 객체를 생성하려면 "생성자"를 먼저 떠올리게됩니다. 정적 팩토리 메서드가 만능일까요? 이펙티브 자바를 읽으셨다면 거기에서 왜 정적 팩토리 메서드를 활용하는지에 대한 케이스를 다시 한번 고민해보세요.

kwonkeonhyeong replied:
정적 팩토리를 왜 사용하는지 그리도 네이밍 컨벤션에 대해서도 학습하고 적절히 사용했는지 다시 고민하여 2단계에 적용해 보도록 하겠습니다.

좋은 리뷰 감사드립니다 😀

jinyoungchoi95 commented on src/main/java/attendance/domain/Command.java:
`Command`와 `AttendanceManager`는 도메인이 맞나요? 미션들을 진행하면서 지켜야하는 가장 큰 우선순위는 좋은 설계, 화려한 기술보다는 미션에 작성된 요구사항이니 미션을 진행하면서 항상 요구사항을 지키고 있는가를 의식해보세요.

kwonkeonhyeong replied:
계층을 다 걷어내는 과정에서 domain과 view 만 남기고 생각해 보자는 생각에서 Command 과 AttendanceManager 또한 콘솔 게임의 전체적인 흐름을 관리하는 도메인이라고 생각하여 도메인 안에 넣어 보았습니다.
해당 부분이 controller의 역할을 맡고 있다고 생각은 하여, step2에서는 남겨주신 리뷰 참고하여 해당 역할들을 하는 객체를 어디에 위치시킬 지 고민해 보겠습니다!

jinyoungchoi95 commented on src/main/java/attendance/domain/crew/AttendanceStatus.java:
```diff
@@ -0,0 +1,9 @@
+package attendance.domain.crew;
+
+public enum AttendanceStatus {
```
껍데기만 있는 enum을 만들기보다 객체를 설계할때부터 어떤 역할을 줄 것인지 명확하게 정의해보세요. enum도 객체입니다.

kwonkeonhyeong replied:
리펙터링 과정에서 해당 enum이 view 와 연결되어 있던 부분을 걷어내니 껍대기만 남아있는 enum이 되었습니다.

아마 다른 객체에 (제가 생각하기엔 timlog)쪽에 너무 많은 책임들이 부여되게 설계가 되어 있어서 이 enum이 껍대기만 남지 않았나 생각합니다.

enum의 쓰임에 대해서 다시 생각해보고 사용해 보도록 하겠습니다 👍

jinyoungchoi95 commented on src/main/java/attendance/domain/crew/comprator/DefaultCrewAttendanceComparator.java:
```diff
@@ -0,0 +1,18 @@
+import attendance.domain.crew.CrewAttendanceStatus;
+import java.util.Comparator;
+
+public class DefaultCrewAttendanceComparator implements CrewAttendanceComparator {
```
중요한 도메인 로직을 가진 클래스로 보이는데 테스트가 필요없을까요?

kwonkeonhyeong replied:
정렬 부분에 대해서는 제적 위험자 반환에 대한 테스트를 진행할 때 정렬된 리스트를 가지고 테스트를 진행해서 해당 테스트 부분에 정렬과 관련되 테스트도 포함되어 있지 않을까? 하는 생각에 테스트를 진행하지 않았습니다.

다시 생각해 보니 comparator에 대한 테스트가 이루어지지 않고 있어서 중요한 도메인 로직을 가진 클래스에 대한 테스트를 확실히 진행해 보겠습니다😂

jinyoungchoi95 commented on src/main/java/attendance/domain/crew/comprator/CrewAttendanceComparator.java:
```diff
@@ -0,0 +1,8 @@
+import attendance.domain.crew.CrewAttendanceStatus;
+import java.util.Comparator;
+
+public interface CrewAttendanceComparator extends Comparator<CrewAttendanceStatus> {
```
현재 추상화를 통해 얻는 이점은 무엇인가요?

kwonkeonhyeong replied:
처음에 설계할 때 정렬하는 방식이 여러가지로 나누어 지는 경우 추상화를 통해서 정렬 기준을 유연하게 갈아낄 수 있지 않을까? 라는 생각에 추상화를 진행해 보았습니다!

jinyoungchoi95 reviewed (approved):
안녕하세요 히포 😄

몇가지 코멘트 남겨두었으니 다음 단계 진행하면서 함께 확인부탁드려요.

궁금하거나 고민이 되는 부분이 있으시다면 언제든 pr 코멘트 또는 dm으로 요청 부탁드립니다.
감사합니다 🙇‍♂️
